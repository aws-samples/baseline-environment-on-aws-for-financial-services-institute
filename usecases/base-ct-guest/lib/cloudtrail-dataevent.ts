import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { aws_s3 as s3 } from 'aws-cdk-lib';
import { aws_cloudtrail as trail } from 'aws-cdk-lib';
import { aws_logs as cwl } from 'aws-cdk-lib';
import { aws_kms as kms } from 'aws-cdk-lib';
import { S3EventSelector } from 'aws-cdk-lib/aws-cloudtrail';

/*
 * CloudTrail S3データイベントの有効化
 * ログはLog Archiveアカウントの集約バケットに保管される
 */

interface CloudTrailStackDataEventProps extends cdk.StackProps {
  cloudTrailBucketName: string; //Log Archiveアカウントに作成したS3バケット名（cdk.jsonで設定）
  targetBuckets: string[]; //CloudTrailデータイベント取得の対象となるS3バケット名を指定（cdk.jsonで指定）
  controlTowerKMSKeyArn: string; //CloudTrail 暗号化キー（ControlTowerの暗号化キーを使用. cdk.jsonで指定）
}

export class CloudTrailDataEvent extends Construct {
  constructor(scope: Construct, id: string, props: CloudTrailStackDataEventProps) {
    super(scope, id);

    //1 find a bucket for cloudtrail
    const cloudTrailBucket = s3.Bucket.fromBucketName(this, 'cloudTrailBucket', props.cloudTrailBucketName);

    //2 get KMS CMK for CloudTrail. it usually use KMS key created in Control Tower
    const ctKmsKey = kms.Key.fromKeyArn(this, 'ct-kmskey', props.controlTowerKMSKeyArn);

    //3 create CloudWatch Logs Group for CloudTrail data logging
    const cloudTrailLogGroup = new cwl.LogGroup(this, 'LogGroup', {
      retention: cwl.RetentionDays.TWO_WEEKS,
    });

    //4 add target S3 buckets for data event to the selector
    const s3eventSelectors: S3EventSelector[] = [];
    for (const bucketName of props.targetBuckets) {
      s3eventSelectors.push({ bucket: s3.Bucket.fromBucketName(this, bucketName, bucketName) });
    }

    //5 Create CloudTrail data event Trails
    const cloudtrail = new trail.Trail(this, 'Default', {
      bucket: cloudTrailBucket,
      enableFileValidation: true,
      includeGlobalServiceEvents: true,
      cloudWatchLogGroup: cloudTrailLogGroup,
      encryptionKey: ctKmsKey,
      sendToCloudWatchLogs: true,
    });

    cloudtrail.addS3EventSelector(s3eventSelectors, {
      includeManagementEvents: false,
      readWriteType: trail.ReadWriteType.ALL,
    });
  }
}
