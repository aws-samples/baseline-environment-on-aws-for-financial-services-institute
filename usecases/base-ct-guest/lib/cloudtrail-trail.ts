import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { PjPrefix } from '../bin/parameter';
import { Environment } from 'aws-cdk-lib';
import { aws_s3 as s3 } from 'aws-cdk-lib';
import { aws_cloudtrail as trail } from 'aws-cdk-lib';
import { aws_logs as cwl } from 'aws-cdk-lib';
import { aws_iam as iam } from 'aws-cdk-lib';
import { Bucket } from './constructs/bleafsi-s3-bucket';
import { KmsKey } from './constructs/bleafsi-kms-key';

/*
 * Control Tower ゲストアカウントでのCloudTrail証跡の作成
 *  Control Tower LandingZone 3.0 からCloudTrailは組織の証跡が有効化され、CloudWatch LogGroupはAuditアカウント側に
 *  作成されるようになったが、ゲストアカウント側でのアラート通知を有効化させるために別途CloudTrail証跡をゲストアカウント側に作成する
 */
export class CloudTrail extends Construct {
  public readonly cloudTrailLogGroup: cwl.LogGroup;

  constructor(scope: Construct, id: string, env?: Environment) {
    super(scope, id);

    //1 CloudTrail 証跡用バケット作成
    // 必要に応じて removalPolicyは変更する。DESTROYはスタック削除時に自動的にバケットも削除される
    const bucketForTrail = new Bucket(this, 'Bucket', {
      bucketName: `${PjPrefix.toLowerCase()}-trail-log-${cdk.Stack.of(this).account}`,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      accessLogsPrefix: 'cloudtraillogs',
      LifecycleRules: [
        {
          enabled: true,
          expiration: cdk.Duration.days(180), //半年
        },
      ],
    });

    //2 CloudTrail 証跡用 KMS CMK 作成
    const kmsCmk = createKmsCmk(this);

    //3 CloudTrail 証跡用 CloudWatch ロググループ作成
    const cloudTrailLogGroup = new cwl.LogGroup(this, 'LogGroup', {
      retention: cwl.RetentionDays.THREE_MONTHS,
      encryptionKey: kmsCmk.key,
    });
    this.cloudTrailLogGroup = cloudTrailLogGroup;

    //4 CloudTrail 証跡作成
    new trail.Trail(this, 'Default', {
      bucket: bucketForTrail.bucket,
      enableFileValidation: true,
      includeGlobalServiceEvents: true,
      cloudWatchLogGroup: cloudTrailLogGroup,
      encryptionKey: kmsCmk.key,
      sendToCloudWatchLogs: true,
    });
  }
}

/////////// private /////////////////
/*
 * CloudTrail証跡用 KMS CMK 作成
 */
function createKmsCmk(scope: Construct): KmsKey {
  // CMK for CloudTrail
  const cloudTrailKey = new KmsKey(scope, 'Kms', {
    description: 'BLEAFSI GovernanceBase: Used for CloudTrail Logs Encryption',
  });

  cloudTrailKey.addToResourcePolicy(
    new iam.PolicyStatement({
      actions: ['kms:GenerateDataKey*'],
      principals: [new iam.ServicePrincipal('cloudtrail.amazonaws.com')],
      resources: ['*'],
      conditions: {
        StringLike: {
          'kms:EncryptionContext:aws:cloudtrail:arn': [`arn:aws:cloudtrail:*:${cdk.Stack.of(scope).account}:trail/*`],
        },
      },
    }),
  );
  cloudTrailKey.addToResourcePolicy(
    new iam.PolicyStatement({
      actions: ['kms:DescribeKey'],
      principals: [new iam.ServicePrincipal('cloudtrail.amazonaws.com')],
      resources: ['*'],
    }),
  );
  cloudTrailKey.addToResourcePolicy(
    new iam.PolicyStatement({
      actions: ['kms:Decrypt', 'kms:ReEncryptFrom'],
      principals: [new iam.AnyPrincipal()],
      resources: ['*'],
      conditions: {
        StringEquals: { 'kms:CallerAccount': `${cdk.Stack.of(scope).account}` },
        StringLike: {
          'kms:EncryptionContext:aws:cloudtrail:arn': [`arn:aws:cloudtrail:*:${cdk.Stack.of(scope).account}:trail/*`],
        },
      },
    }),
  );
  cloudTrailKey.addToResourcePolicy(
    new iam.PolicyStatement({
      actions: ['kms:Encrypt*', 'kms:Decrypt*', 'kms:ReEncrypt*', 'kms:GenerateDataKey*', 'kms:Describe*'],
      principals: [new iam.ServicePrincipal('logs.amazonaws.com')],
      resources: ['*'],
      conditions: {
        ArnEquals: {
          'kms:EncryptionContext:aws:logs:arn': `arn:aws:logs:${cdk.Stack.of(scope).region}:${
            cdk.Stack.of(scope).account
          }:log-group:*`,
        },
      },
    }),
  );

  return cloudTrailKey;
}
