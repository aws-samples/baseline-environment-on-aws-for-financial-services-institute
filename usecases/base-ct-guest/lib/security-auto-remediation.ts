import { Construct } from 'constructs';
import { aws_config as config } from 'aws-cdk-lib';
import { aws_iam as iam } from 'aws-cdk-lib';

/*
 * デフォルトセキュリティグループが閉じている（inbound/outboundを許可していない）ことをチェックするConfigRuleを作成
 * 違反している場合は、SSMにより自動的に修正が行われる
 */
export class SecurityAutoRemediation extends Construct {
  constructor(scope: Construct, id: string) {
    super(scope, id);

    // ConfigRule for Default Security Group is closed  (Same as SecurityHub - need this for auto remediation)
    //
    // See: https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-cis-controls.html#securityhub-cis-controls-4.3
    // See: https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-standards-fsbp-controls.html
    const ruleDefaultSgClosed = new config.ManagedRule(this, 'ConfigRule', {
      identifier: config.ManagedRuleIdentifiers.VPC_DEFAULT_SECURITY_GROUP_CLOSED,
      ruleScope: config.RuleScope.fromResources([config.ResourceType.EC2_SECURITY_GROUP]),
      configRuleName: 'bb-default-security-group-closed',
      description:
        'Checks that the default security group of any Amazon Virtual Private Cloud (VPC) does not allow inbound or outbound traffic. The rule is non-compliant if the default security group has one or more inbound or outbound traffic.',
    });

    // Create an Auto remediation
    new AutoRemediation(this, 'AutoRemediation', ruleDefaultSgClosed);
  }
}

/////////// private Construct /////////////////
class AutoRemediation extends Construct {
  constructor(scope: Construct, id: string, ruleDefaultSgClosed: config.ManagedRule) {
    super(scope, id);
    //1. Role の作成
    const rmDefaultSgRole = new iam.Role(this, 'IamRole', {
      assumedBy: new iam.ServicePrincipal('ssm.amazonaws.com'),
      path: '/',
      managedPolicies: [{ managedPolicyArn: 'arn:aws:iam::aws:policy/service-role/AmazonSSMAutomationRole' }],
    });

    rmDefaultSgRole.addToPolicy(
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: ['ec2:RevokeSecurityGroupIngress', 'ec2:RevokeSecurityGroupEgress', 'ec2:DescribeSecurityGroups'],
        resources: ['*'],
      }),
    );
    rmDefaultSgRole.addToPolicy(
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: ['iam:PassRole'],
        resources: [rmDefaultSgRole.roleArn],
      }),
    );
    rmDefaultSgRole.addToPolicy(
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: ['ssm:StartAutomationExecution'],
        resources: ['arn:aws:ssm:::automation-definition/AWSConfigRemediation-RemoveVPCDefaultSecurityGroupRules'],
      }),
    );

    // 2. SSMを使った自動修復のための ConfigRule Remediation
    // Remediation for Remove VPC Default SecurityGroup Rules  by  SSM Automation
    new config.CfnRemediationConfiguration(this, 'Default', {
      configRuleName: ruleDefaultSgClosed.configRuleName,
      targetType: 'SSM_DOCUMENT',
      targetId: 'AWSConfigRemediation-RemoveVPCDefaultSecurityGroupRules',
      targetVersion: '1',
      parameters: {
        AutomationAssumeRole: {
          StaticValue: {
            Values: [rmDefaultSgRole.roleArn],
          },
        },
        GroupId: {
          ResourceValue: {
            Value: 'RESOURCE_ID',
          },
        },
      },
      automatic: true,
      maximumAutomaticAttempts: 5,
      retryAttemptSeconds: 60,
    });
  }
}
