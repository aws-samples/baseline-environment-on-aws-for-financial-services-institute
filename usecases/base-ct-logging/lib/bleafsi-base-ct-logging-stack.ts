import * as cdk from 'aws-cdk-lib';
import { PjPrefix } from '../bin/parameter';
import { Construct } from 'constructs';
import { aws_s3 as s3 } from 'aws-cdk-lib';
import { aws_iam as iam } from 'aws-cdk-lib';
import { Bucket } from './constructs/bleafsi-s3-bucket';

/*
 * CloudTrail S3 Data Event用の ログ集約S3バケットを作成
 */
export class BaseCTStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: cdk.StackProps) {
    super(scope, id, props);

    // 集約ログ用のバケット作成
    // 必要に応じて removalPolicyは変更する。DESTROYはスタック削除時に自動的にバケットも削除される
    // CloudTrailのログは既に暗号化されているため、バケット自体はSSE-S3暗号化とする
    const logBucket = new Bucket(this, 'Bucket', {
      bucketName: `${PjPrefix.toLowerCase()}-shared-log-${cdk.Stack.of(this).account}`,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      LifecycleRules: [
        {
          enabled: true,
          expiration: cdk.Duration.days(1825), //5年
        },
      ],
    });

    // リソースポリシーの追加
    logBucket.addToResourcePolicy(
      new iam.PolicyStatement({
        sid: 'Restrict Delete* Actions',
        effect: iam.Effect.DENY,
        actions: ['s3:DeleteObject'],
        principals: [new iam.AnyPrincipal()],
        resources: [logBucket.bucket.arnForObjects('*')],
      }),
    );

    logBucket.addToResourcePolicy(
      new iam.PolicyStatement({
        sid: 'AWSBucketPermissions',
        effect: iam.Effect.ALLOW,
        resources: [logBucket.bucket.bucketArn],
        principals: [
          new iam.ServicePrincipal('cloudtrail.amazonaws.com'),
          new iam.ServicePrincipal('config.amazonaws.com'),
        ],
        actions: ['s3:GetBucketAcl', 's3:ListBucket'],
      }),
    );

    logBucket.addToResourcePolicy(
      new iam.PolicyStatement({
        sid: 'AWSBucketDelivery',
        effect: iam.Effect.ALLOW,
        resources: [`${logBucket.bucket.bucketArn}/*/*`],
        principals: [
          new iam.ServicePrincipal('cloudtrail.amazonaws.com'),
          new iam.ServicePrincipal('config.amazonaws.com'),
        ],
        actions: ['s3:PutObject'],
        conditions: {
          StringEquals: {
            's3:x-amz-acl': 'bucket-owner-full-control',
          },
        },
      }),
    );

    //CFn output
    new cdk.CfnOutput(this, 'Shared Log Bucket', {
      value: logBucket.bucket.bucketName,
      description: 'Bucket for CloudTrail log',
    });
  }
}
