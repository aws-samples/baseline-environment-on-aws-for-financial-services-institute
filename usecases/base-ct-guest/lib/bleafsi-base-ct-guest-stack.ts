import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { IamSample } from './iam-sample';
import { SecurityAutoRemediation } from './security-auto-remediation';
import { SecurityAlarm } from './security-alarm';
import { SessionManagerLog } from './session-manager-log';
import { CloudTrail } from './cloudtrail-trail';
import { StackParameter } from '../bin/parameter';
//import { CloudTrailDataEvent } from './cloudtrail-dataevent';

const OSAKA_REGION = 'ap-northeast-3';

/*
 * Control Tower ゲストアカウント用のガバナンスベース構築 Stack
 */
export class BaseCTStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: StackParameter) {
    super(scope, id, props);

    // 1 ゲストアカウントでのCloudTrail証跡の作成
    // Control Tower ランディングゾーン v3.0 からゲストアカウントに CloudWatch ロググループを作成しないように変更されました。
    // また以前の環境からランディングゾーンをv3.0にアップデートした時に CloudTrail の組織レベルのログ記録設定が無効化されていると ゲストアカウント上のロググループを削除します。
    // BLEA for FSI では引き続きゲストアカウント上でのアラーム通知機能を有効にしているため、ゲストアカウント上に CloudWatch ロググループが作成されていない場合は、
    // アラームを通知するために CloudWatch ロググループ をゲストアカウント上に作成します。

    // 下記の条件に当てはまる場合は、block1 のコードのコメントを外し、block2 のコードはコメントアウトして下さい。
    // v3.0より以前のバージョンで Control Tower がゲストアカウント上に作成した CloudWatch ロググループを使用して、新規にロググループを作成しません。
    // [条件]
    //  - Control Tower ランディングゾーンのバージョンを v3.0 にアップデートして、
    //    Control Tower ランディングゾーン設定の AWS CloudTrail 設定で、「組織レベルのログ記録」が 有効になっている

    // ----- block1 ----
    //const cloudTrailLogGroupName = 'aws-controltower/CloudTrailLogs'; //Created by ControlTower for each account
    // ----------------

    // 下記の条件のいずれかに当てはまる場合は、下記の block2 のコードを使用して下さい。
    // - Control Tower ランディングゾーンを v3.0以降から使用している（以前のバージョンからアップデートしていない）
    // - Control Tower ランディングゾーンを v3.0以前から使用している（以前のバージョンからアップデートしている）、
    //   ただし、ランディングゾーン設定の AWS CloudTrail 設定で、「組織レベルのログ記録」を無効化している

    // ----- block2 ----
    const cloudTrail = new CloudTrail(this, `CloudTrail`, props.env);
    const cloudTrailLogGroupName = cloudTrail.cloudTrailLogGroup.logGroupName;
    // -----------------

    //2 CloudWatchメトリクスフィルターによるアラーム、EventBridgeルールを使用したセキュリティアラートの定義
    new SecurityAlarm(this, `SecurityAlarm`, {
      notifyEmail: props.securityNotifyEmail,
      cloudTrailLogGroupName: cloudTrailLogGroupName,
    });

    //3 自動修復設定を持つ Config ルールを設定
    if (props.env?.region != OSAKA_REGION) {
      //2023/4/26 大阪リージョンでは SSM document AWSConfigRemediation-RemoveVPCDefaultSecurityGroupRules がないためデプロイしない
      new SecurityAutoRemediation(this, `SecurityAutoRemediation`);
    }

    //4 SSM セッションマネージャーの監査ログ取得用 S3バケット作成
    const smanagerLog = new SessionManagerLog(this, `SessionManagerLog`);

    //5 管理作業用にIAMポリシー/ロールとIAMグループのサンプルを作成
    new IamSample(this, `IamSample`);

    //6 (オプション) CloudTrail S3データイベントの有効化
    /*new CloudTrailDataEvent(this, `CloudTrail-DataEvent`, {
      cloudTrailBucketName: props.cloudTrailBucketName,
      targetBuckets: props.targetBuckets,
      controlTowerKMSKeyArn: props.controlTowerKMSKeyArn,
    });*/

    //CFn output
    new cdk.CfnOutput(this, 'SSM Session Manager Log Bucket', {
      value: smanagerLog.bucket.bucketName,
      description: 'Bucket for SSM Session Manager Log Bucket',
    });
  }
}
