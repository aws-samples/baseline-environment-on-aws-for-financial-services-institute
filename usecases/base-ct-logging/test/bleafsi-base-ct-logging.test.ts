import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { Annotations, Match } from 'aws-cdk-lib/assertions';
import { AwsSolutionsChecks } from 'cdk-nag';
import { NagSuppressions } from 'cdk-nag';
import { BaseCTStack } from '../lib/bleafsi-base-ct-logging-stack';

const app = new cdk.App();
const pjPrefix = 'BLEAFSI-BASE';

let stack: cdk.Stack;

//jest snapshot check
describe(`${pjPrefix} Compare Snapshot test`, () => {
  test('LoggingAccount Stacks', () => {
    stack = new BaseCTStack(app, `${pjPrefix}-Dev`, {
      env: {
        account: process.env.CDK_DEFAULT_ACCOUNT,
        region: process.env.CDK_DEFAULT_REGION,
      },
    });

    // test with snapshot
    expect(Template.fromStack(stack)).toMatchSnapshot();
  });
});

//cdk-nag check
describe(`${pjPrefix} cdk-nag AwsSolutions Pack`, () => {
  beforeAll(() => {
    //cdk-nag suppression
    // S1: The S3 Bucket has server access logs disabled.
    NagSuppressions.addResourceSuppressionsByPath(stack, '/BLEAFSI-BASE-Dev/Bucket/AccessLogs/Default/Resource', [
      {
        id: 'AwsSolutions-S1',
        reason: 'this s3 bucket is used to store server access log itself',
      },
    ]);

    //cdk-nag check
    cdk.Aspects.of(stack).add(new AwsSolutionsChecks());
  });

  test('No unsupressed Errors', () => {
    const errors1 = Annotations.fromStack(stack).findError('*', Match.stringLikeRegexp('AwsSolutions-.*'));
    try {
      expect(errors1).toHaveLength(0);
      console.log('cdk-nag: no errors for stack ' + stack.stackName);
    } catch (e) {
      console.error(errors1);
      throw e;
    }
  });
});
