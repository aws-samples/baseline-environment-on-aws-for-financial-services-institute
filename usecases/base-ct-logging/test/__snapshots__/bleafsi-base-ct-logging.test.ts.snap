// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BLEAFSI-LogBase Compare Snapshot test LoggingAccount Stacks 1`] = `
{
  "Outputs": {
    "SharedLogBucket": {
      "Description": "Bucket for CloudTrail log",
      "Value": {
        "Ref": "Bucket83908E77",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "Bucket83908E77": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AccessControl": "Private",
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "BucketName": {
          "Fn::Join": [
            "",
            [
              "bleafsi-logbase-shared-log-",
              {
                "Ref": "AWS::AccountId",
              },
            ],
          ],
        },
        "LifecycleConfiguration": {
          "Rules": [
            {
              "ExpirationInDays": 1825,
              "Status": "Enabled",
            },
          ],
        },
        "LoggingConfiguration": {
          "DestinationBucketName": {
            "Ref": "BucketAccessLogs0AA7ED48",
          },
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "Tags": [
          {
            "Key": "aws-cdk:auto-delete-objects",
            "Value": "true",
          },
        ],
        "VersioningConfiguration": {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "BucketAccessLogs0AA7ED48": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AccessControl": "LogDeliveryWrite",
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "BucketName": {
          "Fn::Join": [
            "",
            [
              "bleafsi-logbase-shared-log-",
              {
                "Ref": "AWS::AccountId",
              },
              "-logs",
            ],
          ],
        },
        "LifecycleConfiguration": {
          "Rules": [
            {
              "ExpirationInDays": 2555,
              "Status": "Enabled",
              "Transitions": [
                {
                  "StorageClass": "GLACIER",
                  "TransitionInDays": 90,
                },
              ],
            },
          ],
        },
        "OwnershipControls": {
          "Rules": [
            {
              "ObjectOwnership": "ObjectWriter",
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "Tags": [
          {
            "Key": "aws-cdk:auto-delete-objects",
            "Value": "true",
          },
        ],
        "VersioningConfiguration": {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "BucketAccessLogsAutoDeleteObjectsCustomResourceF1F74F65": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "BucketAccessLogsPolicy6EEF3369",
      ],
      "Properties": {
        "BucketName": {
          "Ref": "BucketAccessLogs0AA7ED48",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3AutoDeleteObjects",
      "UpdateReplacePolicy": "Delete",
    },
    "BucketAccessLogsPolicy6EEF3369": {
      "Properties": {
        "Bucket": {
          "Ref": "BucketAccessLogs0AA7ED48",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "BucketAccessLogs0AA7ED48",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "BucketAccessLogs0AA7ED48",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:PutBucketPolicy",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::GetAtt": [
                    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
                    "Arn",
                  ],
                },
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "BucketAccessLogs0AA7ED48",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "BucketAccessLogs0AA7ED48",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "s3:DeleteObject",
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "BucketAccessLogs0AA7ED48",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
              "Sid": "Restrict Delete* Actions",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "BucketAutoDeleteObjectsCustomResourceBAFD23C2": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "BucketPolicyE9A3008A",
      ],
      "Properties": {
        "BucketName": {
          "Ref": "Bucket83908E77",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3AutoDeleteObjects",
      "UpdateReplacePolicy": "Delete",
    },
    "BucketPolicyE9A3008A": {
      "Properties": {
        "Bucket": {
          "Ref": "Bucket83908E77",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "Bucket83908E77",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "Bucket83908E77",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:PutBucketPolicy",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::GetAtt": [
                    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
                    "Arn",
                  ],
                },
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "Bucket83908E77",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "Bucket83908E77",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "s3:DeleteObject",
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "Bucket83908E77",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
              "Sid": "Restrict Delete* Actions",
            },
            {
              "Action": [
                "s3:GetBucketAcl",
                "s3:ListBucket",
              ],
              "Effect": "Allow",
              "Principal": {
                "Service": [
                  "cloudtrail.amazonaws.com",
                  "config.amazonaws.com",
                ],
              },
              "Resource": {
                "Fn::GetAtt": [
                  "Bucket83908E77",
                  "Arn",
                ],
              },
              "Sid": "AWSBucketPermissions",
            },
            {
              "Action": "s3:PutObject",
              "Condition": {
                "StringEquals": {
                  "s3:x-amz-acl": "bucket-owner-full-control",
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": [
                  "cloudtrail.amazonaws.com",
                  "config.amazonaws.com",
                ],
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "Bucket83908E77",
                        "Arn",
                      ],
                    },
                    "/*/*",
                  ],
                ],
              },
              "Sid": "AWSBucketDelivery",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F": {
      "DependsOn": [
        "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "2eb6a831b107939f63cfebf68e6316e1a40f79fc99cae0fee9b333bac8d29bc4.zip",
        },
        "Description": {
          "Fn::Join": [
            "",
            [
              "Lambda function for auto-deleting objects in ",
              {
                "Ref": "BucketAccessLogs0AA7ED48",
              },
              " S3 bucket.",
            ],
          ],
        },
        "Handler": "index.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
